### PR TITLE
ISPN-12879 Pluggable TransportFactory for the Hot Rod client

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12879

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/package-info.java [hunk 1]
```diff
@@ -98,6 +98,12 @@
  *          <td>The Hot Rod {@link org.infinispan.client.hotrod.configuration.ConfigurationBuilder#version(org.infinispan.client.hotrod.ProtocolVersion) version}.</td>
  *       </tr>
  *       <tr>
+ *          <td><b>infinispan.client.hotrod.transport_factory</b></td>
+ *          <td>String</td>
+ *          <td>{@link org.infinispan.client.hotrod.impl.transport.netty.DefaultTransportFactory}</td>
+ *          <td>The transport factory to use</td>
```
Comment (oraNod) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/package-info.java:104]:
```suggestion
 *          <td>Specifies the transport factory to use.</td>
```

**Comment by tristantarrant** [comment]
Merged @oraNod 's edits as well as some javadoc additions

**Review by oraNod — APPROVED** [review]

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/package-info.java [hunk 2]
```diff
@@ -98,6 +98,12 @@
  *          <td>The Hot Rod {@link org.infinispan.client.hotrod.configuration.ConfigurationBuilder#version(org.infinispan.client.hotrod.ProtocolVersion) version}.</td>
  *       </tr>
  *       <tr>
+ *          <td><b>infinispan.client.hotrod.transport_factory</b></td>
+ *          <td>String</td>
+ *          <td>{@link org.infinispan.client.hotrod.impl.transport.netty.DefaultTransportFactory}</td>
+ *          <td>Specifies the transport factory to use.</td>
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/package-info.java:104]: Probably should mention class name?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/ConfigurationChildBuilder.java [hunk 3]
```diff
@@ -257,6 +258,8 @@
     */
    ConfigurationBuilder transactionTimeout(long timeout, TimeUnit timeUnit);
 
+   ConfigurationBuilder transportFactory(TransportFactory transportFactory);
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/ConfigurationChildBuilder.java:261]: Missing javadoc

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/TransportFactory.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 12.1
+ **/
+public interface TransportFactory {
+   TransportFactory DEFAULT = new DefaultTransportFactory();
+
+   /**
+    * Creates a new {@link SocketChannel}
+    */
+   Class<? extends SocketChannel> socketChannel();
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/TransportFactory.java:22]: Maybe `socketChannelClass` ?

**Review by wburns — COMMENTED** [review]
Just some minor comments.

**Comment by tristantarrant** [comment]
@wburns I've addressed the issues.

**Comment by wburns** [comment]
Checkstyle failure.

**Review by wburns — APPROVED** [review]

**Comment by tristantarrant** [comment]
Damn, fixed it

**Comment by wburns** [comment]
Integrated into master, thanks @tristantarrant !

**Pull request merged**